algorithms/kernel/dtrees/forest/regression/oneapi/cl_kernels/df_batch_regression_kernels.cl

```diff
@@ -0,0 +1,994 @@
+
+        algorithmFPType bestLN        = (algorithmFPType)0;
+        algorithmFPType bestLMean     = (algorithmFPType)0;
+        algorithmFPType bestLSum2Cent = (algorithmFPType)0;
+
+        algorithmFPType mrgRN        = (algorithmFPType)0;
+        algorithmFPType mrgRMean     = (algorithmFPType)0;
+        algorithmFPType mrgRSum2Cent = (algorithmFPType)0;
+
+        algorithmFPType mrgLN        = (algorithmFPType)0;
+        algorithmFPType mrgLMean     = (algorithmFPType)0;
+        algorithmFPType mrgLSum2Cent = (algorithmFPType)0;
+
+        int totalBins = 0;
+
+        // totalBins is calculated by each subgroup
+        for (int featIdx = sub_group_local_id; featIdx < nSelectedFeatures; featIdx += sub_group_size)
+        {
+            int featId = selectedFeatures[nodeId * nSelectedFeatures + featIdx];
+            int nBins  = binOffsets[featId + 1] - binOffsets[featId];
+            totalBins += sub_group_reduce_add(nBins);
+        }
+        totalBins = sub_group_broadcast(totalBins, 0);
+
+        int currFtrIdx  = 0;
+        int featId      = selectedFeatures[nodeId * nSelectedFeatures + currFtrIdx];
+        int binId       = 0;
+        int currFtrBins = binOffsets[featId + 1] - binOffsets[featId];
+        int passedBins  = 0;
+
+        for (int i = local_id; i < totalBins; i += local_size)
+        {
+            while (i >= passedBins + currFtrBins)
+            {
+                passedBins += currFtrBins;
+                currFtrIdx++;
+                featId      = selectedFeatures[nodeId * nSelectedFeatures + currFtrIdx];
+                currFtrBins = binOffsets[featId + 1] - binOffsets[featId];
+            }
+            binId = i - passedBins;
+
+            mrgN        = (algorithmFPType)0;
+            mrgMean     = (algorithmFPType)0;
+            mrgSum2Cent = (algorithmFPType)0;
+
+            mrgRN        = (algorithmFPType)0;
+            mrgRMean     = (algorithmFPType)0;
+            mrgRSum2Cent = (algorithmFPType)0;
+
+            mrgLN        = (algorithmFPType)0;
+            mrgLMean     = (algorithmFPType)0;
+            mrgLSum2Cent = (algorithmFPType)0;
+
+            for (int row = 0; row < nRows; row++)
+            {
+                int id  = treeOrder[rowsOffset + row];
+                int bin = data[id * nFeatures + featId];
+
+                if (bin <= binId)
+                {
+                    mergeValToStat(response[id], &mrgLN, &mrgLMean, &mrgLSum2Cent);
+                }
+                else
+                {
+                    mergeValToStat(response[id], &mrgRN, &mrgRMean, &mrgRSum2Cent);
+                }
+            }
+
+            mrgN        = mrgLN;
+            mrgMean     = mrgLMean;
+            mrgSum2Cent = mrgLSum2Cent;
+            mergeStat(mrgRN, mrgRMean, mrgRSum2Cent, &mrgN, &mrgMean, &mrgSum2Cent);
+
+            algorithmFPType impDec = mrgSum2Cent - (mrgLSum2Cent + mrgRSum2Cent);
+
+            if (mrgSum2Cent / mrgN < impurityThreshold)
```

Comment:
Is this safe to exit from kernel under if condition?
I mean, is the result of this condition equal on all threads?

Reply:
yes, due to it is merged centered sum for the node and because of this condition will be true for all threads targeted for the node

Reply:
i merged this check with next one and removed "return"